File: game/scripts/vscripts/legacy/abilities/heroes/invoker/invoker_extra/invoker_walk_custom.lua
invoker_walk_custom = class({})
LinkLuaModifier("modifier_invoker_walk_custom", "abilities/heroes/invoker/invoker_extra/modifier_invoker_walk_custom", LUA_MODIFIER_MOTION_NONE)

function invoker_walk_custom:GetCastAnimationCustom()		return ACT_DOTA_CAST_GHOST_WALK end
function invoker_walk_custom:GetPlaybackRateOverride() 	    return 1.2 end
function invoker_walk_custom:GetCastPointSpeed() 			return 100 end

function invoker_walk_custom:OnSpellStart()
    local caster = self:GetCaster()
    local duration = self:GetSpecialValueFor("duration")
    caster:AddNewModifier(caster, self, "modifier_invoker_walk_custom", { duration = duration })
    
    EFX("particles/units/heroes/hero_invoker/invoker_ghost_walk.vpcf", PATTACH_ABSORIGIN_FOLLOW, caster, {
        release = true
    })
	EmitSoundOn("Hero_Invoker.GhostWalk", caster)
end

if IsClient() then require("wrappers/abilities") end
Abilities.Castpoint(invoker_walk_custom)

File: src/vscripts/abilities/heroes/invoker/invocations/invoker_walk_custom.ts
import { registerAbility, registerModifier } from "../../../../lib/dota_ts_adapter";
import { CustomAbility } from "../../../framework/custom_ability";
import { CustomModifier } from "../../../framework/custom_modifier";

@registerAbility("invoker_walk_custom")
class InvokerWalk extends CustomAbility {
    GetAnimation() {
        return GameActivity.DOTA_CAST_GHOST_WALK;
    }

    GetPlaybackRateOverride() {
        return 1.2;
    }

    GetCastingCrawl() {
        return 100;
    }

    OnSpellStart() {
        ModifierInvokerWalk.apply(this.caster, this.caster, this, { duration: this.GetSpecialValueFor("duration") });

        EFX("particles/units/heroes/hero_invoker/invoker_ghost_walk.vpcf", ParticleAttachment.ABSORIGIN_FOLLOW, this.caster, {
            release: true
        });
        EmitSoundOn("Hero_Invoker.GhostWalk", this.caster);
    }
}

@registerModifier({ customNameForI18n: "modifier_invoker_walk_custom" })
class ModifierInvokerWalk extends CustomModifier {
    IsHidden() {
        return false;
    }

    IsDebuff() {
        return false;
    }

    IsPurgable() {
        return true;
    }

    OnDestroy() {
        if (IsServer()) {
            this.ability.StartCooldown(this.ability.GetCooldown(0));
        }
    }

    DeclareFunctions() {
        return [ModifierFunction.INVISIBILITY_LEVEL, ModifierFunction.MOVESPEED_BONUS_PERCENTAGE];
    }

    GetModifierInvisibilityLevel() {
        return 2;
    }

    GetModifierMoveSpeedBonus_Percentage() {
        return this.Value("speed_buff_pct");
    }

    CheckState() {
        return {
            [ModifierState.INVISIBLE]: true,
            [ModifierState.TRUESIGHT_IMMUNE]: false
        };
    }

    // function modifier_invoker_walk_custom:GetStatusLabel() return "Ghost Walk" }
    // function modifier_invoker_walk_custom:GetStatusPriority() return 1 }
    // function modifier_invoker_walk_custom:GetStatusStyle() return "GhostWalk" }
}
// if(IsClient() ) { require("wrappers/modifiers") }
// Modifiers.Status(modifier_invoker_walk_custom)
